Register with Mesos and launch a real task

timcharper opened the pull request:
JIRA Issues: DCOS-47474

Commit:
Use protos rather than mocked data

Commit:
Initial resource matching logic

Commit:
Working with matchers

Commit:
Real task is launched

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/ProtoBuilders.scala:
```diff
@@ -0,0 +1,201 @@
+import com.google.protobuf.ByteString
+import org.apache.mesos.v1.{Protos => Mesos}
+
+private[usi] object ProtoBuilders {
```
This seems similar to [ProtosHelper](https://github.com/mesosphere/usi/blob/master/examples/simple-hello-world/src/main/scala/com/mesosphere/mesos/examples/ProtosHelper.scala). Should we consolidate them?

zen-dog replied:
I don't think so. The whole idea of these examples is to show the reader what he/she gets by taking core (e.g. not having to build Protobufs). It's fine to have this duplicated imho

jeschkies commented on core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala:
```diff
@@ -0,0 +1,46 @@
+      .queue[SpecEvent](16, OverflowStrategy.fail)
+      .via(client.schedulerFlow)
+      .toMat(Sink.queue())(Keep.both)
+      .run
```
This is interesting. @zen-dog and I were wondering how one would use the flow. What do you think of returning `(Source[SpecEvent], Sink[StateEvent])` instead of `Flow[SpecEvent, StateEvent]` as we do now?

timcharper replied:
It's something we can explore; there is the thing where the Source and the Sink should share the same materialization as they're backed by the same scheduler logic component; but we could split apart a flow by using Sink.asPublisher etc.

jeschkies replied:
Ah, that's a good argument. I'd love to have the separation though since I can think of cases when I'd like to have pulling and pushing blocking in a thread on my side when the orchestration code cannot be async.

timcharper replied:
Yes; and, you could also have a graph on the other side

zen-dog commented on core/src/test/scala/com/mesosphere/usi/core/SchedulerTest.scala:
```diff
@@ -3,27 +3,28 @@ package com.mesosphere.usi.core
+import com.mesosphere.mesos.client.MesosCalls
+import com.mesosphere.usi.core.launching.SimpleShellCommandInfoGenerator
+import com.mesosphere.usi.core.matching.ScalarResourceRequirement
+import com.mesosphere.usi.core.models._
```
Oh no, you did not! :P

timcharper replied:
Organize imports did it I swear

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -173,51 +255,67 @@ private[core] object SchedulerLogic {
+            if (podSpec.goal == Goal.Terminal) {
+              taskIdsFor(podSpec).foldLeft(initialIntents) { (effects, taskId) =>
+                effects.withMesosCall(
+                  mesosCallFactory.newKill(taskId.asProto, state.podRecords.get(podSpec.id).map(_.agentId.asProto), None))
```
is it cool as we're initiating kill for any change? E.g. the change is emitted even on a snapshot. Or we're going to tackle this later?

timcharper replied:
Yeah, i think it is fine for now. We'll want a more robust kill mechanism with retry I think in the future perhaps? We should add a JIRA

timcharper replied:
https://jira.mesosphere.com/browse/DCOS-48513

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/PodStatus.scala:
```diff
@@ -1,5 +1,7 @@
 package com.mesosphere.usi.core.models
 
+import org.apache.mesos.v1.{Protos => Mesos}
```
Where is that used?

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/ResourceRequirement.scala:
```diff
@@ -0,0 +1,9 @@
+
+import org.apache.mesos.v1.{Protos => Mesos}
+
+trait ResourceRequirement {
```
Could you add docs since this is part of our protocol?

timcharper replied:
✅

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -173,51 +255,67 @@ private[core] object SchedulerLogic {
+
+      case UpdateEvent(updateEvent) =>
+        val taskStatus: Mesos.TaskStatus = updateEvent.getStatus
+        val taskId = TaskId(taskStatus.getTaskId.getValue)
```
can we make UpdateEvent a case class and have these (taskId, podId) there as a vals/fields?

timcharper replied:
Perhaps? I am trying to limit the amount of special containers we use for Protobufs, and instead use helper methods to read them.

timcharper replied:
I misread what you wrote :) Yes this is a good idea.

Commit:
Renames, push snapshot -> updates type interface through

Commit:
doc ResourceRequirement

Commit:
Enrich the matchers for MesosEvent

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/ResourceUtil.scala:
```diff
@@ -0,0 +1,218 @@
+import scala.collection.JavaConverters._
+import scala.util.control.NonFatal
+
+object ResourceUtil extends StrictLogging {
```
note for reviewer: this code seems to be basically copy&paste from Marathon. I'll follow up with migrating the tests as well

timcharper replied:
I didn't see this, I already migrated the tests.

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -43,44 +50,84 @@ import scala.concurrent.Future
+
+  case class MesosConnection(mesosHostName: String, mesosCallFactory: MesosCalls, frameworkId: FrameworkID, schedulerFlow: Flow[SpecInput, StateOutput, NotUsed])
+
+  // TODO - provide abstraction for FrameworkID persistence, recovery, etc.; we'll want to exclude the frameworkId
```
should we capture this in JIRA?

alenkacz replied:
also for the comment below

zen-dog replied:
how about injecting the complete `MesosClient` here? This way `core` can be easily tested with a Mesos mock

zen-dog replied:
Discussed

timcharper replied:
✅  I just rolled this stuff back and made it so the MesosClient is passed in

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -43,44 +50,84 @@ import scala.concurrent.Future
  *                                     +----------------------+
  */
 object Scheduler {
```
what do you think about PodScheduler?

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -43,44 +50,84 @@ import scala.concurrent.Future
+  def unconnectedGraph(mesosCallFactory: MesosCalls): BidiFlow[SpecInput, StateOutput, MesosEvent, MesosCall, NotUsed] = {
     BidiFlow.fromGraph {
-      GraphDSL.create(new SchedulerLogicGraph) { implicit builder => (schedulerLogic) =>
+      GraphDSL.create(new SchedulerLogicGraph(mesosCallFactory)) { implicit builder => (schedulerLogic) =>
```
I would go for `SchedulerGraph`

timcharper replied:
We need a name for the component that specifically refers to what is currently called SchedulerLogic, then. Scheduler includes persistence component as we discussed yesterday.

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/logic/MesosEventsLogic.scala:
```diff
@@ -0,0 +1,144 @@
+    }
+  }
+
+  def matchOffer(offer: Mesos.Offer, specs: Iterable[PodSpec]): (Set[PodId], SchedulerEventsBuilder) = {
```
should be private

timcharper replied:
✅

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/logic/SpecificationLogic.scala:
```diff
@@ -0,0 +1,93 @@
+  *
+  * TODO separate the Mesos and the specification logic.
+  */
+private[core] class SpecificationLogic(mesosCallFactory: MesosCalls) extends StrictLogging {
```
Specification -> Spec

timcharper replied:
✅

alenkacz reviewed (commented):
After about half-done I have to take a step back and try some stuff in IDEA to see if we can do better in some places. Then I'll come back to continue the review :)

zen-dog commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunSpec.scala:
```diff
@@ -2,6 +2,5 @@ package com.mesosphere.usi.core.models
-  * @param cpus number of cpus requested by the pod
   */
-case class RunSpec(cpus: Double)
+case class RunSpec(resourceRequirements: Seq[ResourceRequirement], commandBuilder: CommandInfoGenerator)
```
I don't know how I feel about `commandBuilder` in the `RunSpec`. I guess it's here because we wanted to keep it flexible for the orchestrators? But on the other hand, orchestrators should be primarily about business logic and not about building Mesos protobufs correctly. Maybe we could _make it extendable later_ similar to how plugins in Marathon _can but not must_ override `TaskInfo` and `TaskGroupInfo`?

alenkacz replied:
We already discussed on slack but just for reference. The usage of this is  `RunSpec(SimpleShellCommandInfoGenerator("echo 'abc'"))`. I would personally prefer something like `RunSpec(ShellCommand("cmd"))` but we can iterate on that

timcharper replied:
Yeah, I'm not exactly sure at the moment. 

https://jira.mesosphere.com/browse/DCOS-48503
https://jira.mesosphere.com/browse/DCOS-47481

timcharper replied:
Replaced with string shell command for now, with call out to above JIRAs

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/ResourceUtil.scala:
```diff
@@ -0,0 +1,218 @@
+      resource.hasDisk && resource.getDisk.hasSource &&
+        (resource.getDisk.getSource.getType == Source.Type.MOUNT)
+
+    // TODO(jdef) would be nice to use fixed precision like Mesos does for scalar math
```
`// TODO(jdef)`? 🤣 Is James aware?

alenkacz replied:
yeah it made me laugh as well :D

timcharper replied:
lol copied from Marathon

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -43,44 +50,84 @@ import scala.concurrent.Future
+  case class MesosConnection(mesosHostName: String, mesosCallFactory: MesosCalls, frameworkId: FrameworkID, schedulerFlow: Flow[SpecInput, StateOutput, NotUsed])
+
+  // TODO - provide abstraction for FrameworkID persistence, recovery, etc.; we'll want to exclude the frameworkId
+  // TODO - This layer should provide a cancellation mechanism and an auto-reconnect mechanism
```
Cancellation - yes, auto-reconnect - I don't think so (especially when thinking about HA frameworks).

timcharper replied:
No? We should crash on Mesos connection loss?

zen-dog replied:
Signal that to the orchestration layer and let it decide, I'd say

zen-dog replied:
Discussed

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -43,44 +50,84 @@ import scala.concurrent.Future
+    * proper implementation path. A SpecSnapshot should be the very first thing that the USI Scheduler receives from the
+    * Framework implementation.
+    *
+    * However, for convenience in working with streams, internally we deal with a single stream of SpecEvents.
```
I imagined that scheduler would start Mesos task reconciliation after receiving the `SpecSnapshot` since this is a good point in time when all the reconciled tasks should be ideally found in the snapshot. Wdyt?

alenkacz replied:
I believe it should be the case as well. But I think reconciliation is out of scope for now

timcharper replied:
Yes, https://jira.mesosphere.com/browse/DCOS-47476

zen-dog replied:
`//TODO:  DCOS-48503`

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -43,44 +50,84 @@ import scala.concurrent.Future
+      case c: StateUpdated => c
+      case _ => ???
+    }
+    (StateSnapshot.empty, stateUpdates)
```
The other way around (see my comment above), I'm not sure what difference does it make for the orchestration layer to receive `StateSnapshot` as opposed to a stream of `StateUpdated`.

alenkacz replied:
I believe it's just enforcing the order of Snapshot always being first. You need the snapshot to finish initialization of your orchestration layer

zen-dog replied:
Discussed

alenkacz reviewed (approved):
I've submitted few patches in https://github.com/mesosphere/usi/pull/41 with that getting in I am fine with this. There's still a lot to polish but I would rather iterate on that while we evolve core + hello-world side by side.

The only blocker for me is ResourceUtil tests. I think those should be ported here.

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicHandler.scala:
```diff
@@ -0,0 +1,146 @@
+  * ## Intents and Events
+  *
+  * In the SchedulerLogic code, we'll use the word intents and events. Intents are things not yet applied, and should
+  * be. Events are things that were applied and we're notifying you about.
```
...and should be...?

timcharper replied:
Yeah, this seems correct to me. What's unclear?

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicHandler.scala:
```diff
@@ -0,0 +1,146 @@
+  *      v
+  * (beginning of frame)
+  *   apply pod launch logic:
+  *     emit Mesos accept offer call for matched pending launch pods
```
```suggestion
  *     create an intent for Mesos accept offer call for matched pending launch pods
```

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicHandler.scala:
```diff
@@ -0,0 +1,146 @@
+  * In the SchedulerLogic, each event is processed (either a Specification updated event, or a Mesos event) in what we
+  * call a "frame". During the processing of a single event (such as an offer), the SchedulerLogic will be updated
+  * through the application of several functions. The result of that event will lead to the emission of several
+  * stateEvents and Mesos intents, which will be accumulated and emitted via a data structure we call the FrameResult.
```
Nice docs 👏

alenkacz committed:
Minor fixes on top of launch-real-mesos-task branch (#41)

* Minor fixes

* Added by mistake

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SpecState.scala:
```diff
@@ -0,0 +1,8 @@
+package com.mesosphere.usi.core
+import com.mesosphere.usi.core.models.{PodId, PodSpec}
+
+case class SpecState(podSpecs: Map[PodId, PodSpec])
```
nit: do we need a case class wrapper? can it be a
```suggestion
type SpecState = Map[PodId, PodSpec]
```

timcharper replied:
We'll have other specs in here, reservation specs

zen-dog commented on core/src/test/scala/com/mesosphere/usi/core/helpers/FakeMesos.scala:
```diff
@@ -0,0 +1,61 @@
+import org.apache.mesos.v1.scheduler.Protos.{Call => MesosCall, Event => MesosEvent}
+import scala.collection.JavaConverters._
+
+object FakeMesos {
```
It looks like this class is to stay. Should we call it `MesosMock`?

timcharper replied:
¯\\\_(ツ)\_/¯ push a change to rename it to whatever you like

zen-dog commented on core/src/test/scala/com/mesosphere/usi/core/helpers/FakeMesos.scala:
```diff
@@ -0,0 +1,61 @@
+object FakeMesos {
+  import com.mesosphere.usi.core.protos.ProtoBuilders._
+  import com.mesosphere.usi.core.protos.ProtoConversions._
+  val fakeAgentId = AgentId("fake-agent")
```
and `mockAgentId`? and `mock-agent`? and `mockMesosLoggingFlow`? and... you get the idea ;)

timcharper replied:
lol maybe later hahaha lets merge this already

zen-dog commented on core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala:
```diff
@@ -0,0 +1,51 @@
+
+    val client = Scheduler.connect(settings, frameworkInfo).futureValue
+
+    val podId = PodId("pod-1")
```
Can we please not use the `pod1` and `app1` names? Please keep them unique and readable (take Marathon ITs as an example)

timcharper replied:
✅

Commit:
Add ResourceUtilTests

Update ResourceUtil to detail only with reservations!

Commit:
Remove psuedo generic commandInfoGenerator to make it clearer that we have not yet planned this part

Commit:
Remove MesosClient connection logic from Scheduler

Commit:
Use a fancy pod name

Commit:
Update core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicHandler.scala

Co-Authored-By: timcharper <tharper@mesosphere.io>

zen-dog committed:
Renamed `FakeMesos` to `MesosMock`

also:
- removed `CommandInfoGenerator` since it wasn't used
- run `gradle build` to get formatting right

zen-dog commented:
I renamed the `FakeMesos` and pushed it however there is still a problem when building the project:

```
* What went wrong:
Execution failed for task ':core-models:compileScala'.
> class org.codehaus.groovy.runtime.GStringImpl cannot be cast to class java.lang.String (org.codehaus.groovy.runtime.GStringImpl is in unnamed module of loader java.net.URLClassLoader @34ce8af7; java.lang.String is in module java.base of loader 'bootstrap')
```

seems that the introduced silencer plugin exposes a groovy string somewhere and it is not properly converted to `java.lang.String`, however, I didn't have time to dig deeper.

zen-dog reviewed (approved):
Lgtm, so we could merge and move on. However, I'd like you to remove the `TODO`s from the scheduler code (there are 14 of those already) - let it fail explicitly for things that aren't implemented and/or create a JIRA ticket to track what needs to be done.

Commit:
Go over TODOs and remove, clarify, add a JIRA, etc

Commit:
clarify todo

Commit:
revert silencer compiler plugin addition

ended up not needing it right now; we can add it later when needed. It broke things

timcharper commented:
I reverted the silencer plugin. After refactoring the ResourceUtil to specifically target multi-role reservations and allocationInfo, the failures are removed.

Commit:
formatting
